File: notification-engine-mern/backend/src/controllers/NotificationController.ts
import { Request, Response } from 'express';
import { Notification } from '../models/Notification';
import { DedupeService } from '../services/DedupeService';
import { AlertFatigueService } from '../services/AlertFatigueService';
import { OpenAiService } from '../services/OpenAiService';
import crypto from 'crypto';

export class NotificationController {
    public static async receiveEvent(req: Request, res: Response) {
        try {
            const payload = req.body;

            if (!payload.user_id || !payload.event_type || !payload.message || !payload.source) {
                return res.status(400).json({ error: 'Missing required fields' });
            }

            // 1. Deduplication
            let dedupeKey = payload.dedupe_key;
            if (!dedupeKey) {
                dedupeKey = crypto.createHash('sha256').update(`${payload.user_id}-${payload.event_type}-${payload.message}`).digest('hex');
            }

            const isExact = await DedupeService.isExactDuplicate(dedupeKey);
            if (isExact) {
                return res.status(202).json({ message: 'Accepted (Exact Duplicate Ignored)' });
            }

            const isNear = await DedupeService.isNearDuplicate(payload.user_id, payload.message);
            if (isNear) {
                return res.status(202).json({ message: 'Accepted (Near Duplicate Ignored)' });
            }

            // 2. Alert Fatigue
            const isFatigued = await AlertFatigueService.isFatigued(payload.user_id);
            if (isFatigued) {
                // Optionally save as 'NEVER' or just ignore
                payload.status = 'NEVER';
                payload.priority_hint = 'Fatigue';
            }

            // 3. Save as PENDING
            const notification = new Notification({
                user_id: payload.user_id,
                event_type: payload.event_type,
                message: payload.message,
                source: payload.source,
                priority_hint: payload.priority_hint,
                channel: payload.channel,
                metadata: payload.metadata,
                dedupe_key: dedupeKey,
                status: payload.status || 'PENDING',
                expires_at: payload.expires_at,
                timestamp: payload.timestamp || new Date()
            });

            await notification.save();

            // 4. Trigger Async Classification if not already bypassed by Fatigue
            if (notification.status === 'PENDING') {
                // Fire and forget
                OpenAiService.processAsynchronously(notification._id.toString());
            }

            // 5. Immediate Return
            return res.status(202).json({ message: 'Accepted', id: notification._id });

        } catch (error: any) {
            console.error('Ingestion Error:', error);
            return res.status(500).json({ error: 'Internal Server Error' });
        }
    }

    // Debug/UI endpoint
    public static async getNotifications(req: Request, res: Response) {
        try {
            const limit = parseInt(req.query.limit as string) || 50;
            const status = req.query.status;

            let query = {};
            if (status) query = { status };

            const notifs = await Notification.find(query).sort({ timestamp: -1 }).limit(limit);
            res.json(notifs);
        } catch (err) {
            res.status(500).json({ error: 'Failed to fetch' });
        }
    }
}

File: notification-engine-spring/backend/src/main/java/com/nexus/notification/controller/NotificationController.java
package com.nexus.notification.controller;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.nexus.notification.model.Notification;
import com.nexus.notification.repository.NotificationRepository;
import com.nexus.notification.service.AlertFatigueService;
import com.nexus.notification.service.DedupeService;
import com.nexus.notification.service.OpenAiService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HexFormat;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/notifications")
@RequiredArgsConstructor
@CrossOrigin(origins = "*") // Allow frontend
public class NotificationController {

    private final NotificationRepository notificationRepository;
    private final DedupeService dedupeService;
    private final AlertFatigueService alertFatigueService;
    private final OpenAiService openAiService;
    private final ObjectMapper objectMapper;

    @PostMapping("/events")
    public ResponseEntity<?> receiveEvent(@RequestBody Map<String, Object> payload)
            throws NoSuchAlgorithmException, JsonProcessingException {
        if (!payload.containsKey("user_id") || !payload.containsKey("event_type") || !payload.containsKey("message")) {
            return ResponseEntity.badRequest().body(Map.of("error", "Missing required fields"));
        }

        String userId = (String) payload.get("user_id");
        String eventType = (String) payload.get("event_type");
        String message = (String) payload.get("message");

        String dedupeKey = (String) payload.get("dedupe_key");
        if (dedupeKey == null) {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest((userId + "-" + eventType + "-" + message).getBytes(StandardCharsets.UTF_8));
            dedupeKey = HexFormat.of().formatHex(hash);
        }

        if (dedupeService.isExactDuplicate(dedupeKey)) {
            return ResponseEntity.status(HttpStatus.ACCEPTED).body(Map.of("message", "Accepted (Exact Duplicate)"));
        }

        if (dedupeService.isNearDuplicate(userId, message)) {
            return ResponseEntity.status(HttpStatus.ACCEPTED).body(Map.of("message", "Accepted (Near Duplicate)"));
        }

        Notification notification = new Notification();
        notification.setUserId(userId);
        notification.setEventType(eventType);
        notification.setMessage(message);
        notification.setSource((String) payload.getOrDefault("source", "api"));
        notification.setPriorityHint((String) payload.get("priority_hint"));
        notification.setChannel((String) payload.get("channel"));
        notification.setDedupeKey(dedupeKey);

        if (payload.containsKey("metadata")) {
            notification.setMetadata(objectMapper.writeValueAsString(payload.get("metadata")));
        }

        if (alertFatigueService.isFatigued(userId)) {
            notification.setStatus("NEVER");
            notification.setPriorityHint("Fatigue Triggered");
        }

        notification = notificationRepository.save(notification);

        if ("PENDING".equals(notification.getStatus())) {
            // Fire and forget Async
            openAiService.processAsynchronously(notification.getId(), payload);
        }

        return ResponseEntity.status(HttpStatus.ACCEPTED)
                .body(Map.of("message", "Accepted", "id", notification.getId()));
    }

    @GetMapping
    public ResponseEntity<List<Notification>> listNotifications(@RequestParam(required = false) String status) {
        if (status != null) {
            return ResponseEntity.ok(notificationRepository.findTop50ByStatusOrderByTimestampDesc(status));
        }
        return ResponseEntity.ok(notificationRepository.findTop50ByOrderByTimestampDesc());
    }
}
